docs/source/cb3.rst

```diff
@@ -169,38 +170,94 @@ Global pinning
 **Summary:**
 
 - Previously we pinned packages using the syntax ``- zlib {{ CONDA_ZLIB }}*``
-- Instead, we should now pin packages with the syntax ``- zlib {{ zlib }}``.
+  in both the ``build`` and ``run`` dependencies.
+- Instead, we should now just specify package names in the ``host`` section,
+  e.g., as simply ``zlib``. They are pinned automatically. No need to add them
+  to run dependencies, as they will be added automatically.
```

Comment:
> No need to add them to run dependencies, as they will be added automatically.

This is only true if the package, e.g., `zlib`, is built with `run_exports` **and** it is ensured that the package build with that `run_exports` is used in `host`. For a package like `zlib`, which has few (=only one) dependencies the following is unlikely to happen, but imagine the following case:
```yaml
# conda_build_config.yaml
upstream: 1.0
pin_run_as_build:
  upstream:
    max_pin: x.x
```
```yaml
package:
  name: downstream
requirements:
  host:
    - upstream
    - other-package <2.0
```
```yaml
package:
  name: upstream
  version: 1.0
build:
  number: 0
requirements:
  run: []
```
```yaml
package:
  name: upstream
  version: 1.0
build:
  number: 1
  run_exports:
    - {{ pin_subpackage('upstream', max_pin='x.x') }}
  requirements:
    run:
      - other-package >=2.0
```

Even though we have a build of `upstream 1.0` that provides `run_exports`, `conda`'s solver will pick up the older build without `run_exports` due to a dependency conflict (and since `conda` doesn't always use the builds with the highest numbers :unamused:). Hence, `downstream` will not have that dependency added to `requirements/run`!
If, as a backup to those cases, we have `pin_run_as_build` and a plain `- upstream` dependency in `requirements/run`, we can avoid that case.
Thus, I think we should recommend
```yaml
package:
  name: downstream
requirements:
  host:
    - upstream
    - other-package <2.0
  run:
    - upstream
```
over
```yaml
package:
  name: downstream
requirements:
  host:
    - upstream
    - other-package <2.0
```

Reply:
Also note that `conda-forge-pinning` only specifies `zlib: 1.2`. Hence, we can create a real world example:
```yaml
package:
  name: downstream
requirements:
  host:
    - bwa 0.7.16
    - zlib
```
=> `bwa 0.7.16` requires `zlib 1.2.8.*` and hence prevents `zlib 1.2.11` (which has \[on `defaults`\] or will have \[on `conda-forge`\] `run_exports`) to be picked up and instead will use `zlib 1.2.8`. Thus, without `requirements: run: ['zlib']` we won't get the `zlib` run requirement.

Reply:
I agree, packages that are needed as run dependencies should always be added to the run section. This also improves readability, as you immediately see what will be used at runtime.

Reply:
Imho we should do what conda-forge is doing here. They are currently integrating the run_dependencies and updating the pinnings accordingly. I don't follow the `bwa` example, as we need to rebuild against bwa either way, against new compiler and against new zlib. Only if we do this, we remove the run dependency. Or do you mean that bwa as tool needs 1.2.8 of zlib? In this case we could 1) rebuild old zlib, 2) patch bwa or 3) make an exception.
While I agree with Johannes that this magic is bad for readability, I see the libgcc (strong run_export) mess we had over the years finally solved with this and I prefer adapting conda-forge strategy for consistency.

Please note that it is not yet entirely clear what conda-forge will be doing and these discussions are all ongoing so if we really disagree we should start discussing in the conda-forge channel.

Reply:
Read the `bwa 0.7.16` as "some random, possibly transitively, additional dependency that causes `conda` to pick up some older `zlib` version without `run_exports`".
I also agree with the readability aspect but this is not the reason why I think the explicit `run` requirements are needed. Only relying on the implicit `run_exports` and `conda` not picking up older versions or builds (within the version constraints given by the package keys in the `conda_build_config.yaml` files) is too brittle. @bgruening, can you point me to some conda-forge discussion where the `run_exports`-only strategy is proposed or discussed? I assume they are aware of the issue I'm pointing at, but if not completely, I'd like to chime in.

Reply:
xrefs from a linked Gitter conversation:
https://gitter.im/conda-forge/conda-forge.github.io?at=5b1006bfb6eece791dca9b6e
https://gitter.im/conda-forge/conda-forge.github.io?at=5b1014c8e26c847ac8a799ec
https://gitter.im/conda-forge/conda-forge.github.io?at=5b10221ba7abc8692efc7b69